atlasdb-client/src/main/java/com/palantir/atlasdb/cache/OffHeapTimestampCache.java

```diff
@@ -0,0 +1,156 @@
+
+        CacheDescriptor previous = cacheDescriptor.getAndUpdate(prev -> proposedCacheDescriptor);
+        if (previous != null) {
+            persistentTimestampStore.dropNamespace(previous.storeNamespace());
+        }
+    }
+
+
+    @Override
+    public void putAlreadyCommittedTransaction(Long startTimestamp, Long commitTimestamp) {
+        if (concurrentHashMap.putIfAbsent(startTimestamp, commitTimestamp) != null) {
+            return;
+        }
+        Futures.getUnchecked(autobatcher.apply(Maps.immutableEntry(startTimestamp, commitTimestamp)));
+    }
+
+    @Nullable
+    @Override
+    public Long getCommitTimestampIfPresent(Long startTimestamp) {
+        Long value = concurrentHashMap.get(startTimestamp);
+        if (value != null) {
+            return value;
+        }
+
+        return persistentTimestampStore.get(cacheDescriptor.get().storeNamespace(), startTimestamp);
+    }
+
+    private static CacheDescriptor constructCacheProposal(PersistentTimestampStore persistentTimestampStore) {
+        StoreNamespace proposal = persistentTimestampStore.createNamespace(TIMESTAMP_CACHE_NAMESPACE);
+        return ImmutableCacheDescriptor.builder()
+                .currentSize(new AtomicInteger())
+                .storeNamespace(proposal)
+                .build();
+    }
+
+    private static class WriteBatcher
+            implements CoalescingRequestFunction<Map.Entry<Long, Long>, Map.Entry<Long, Long>> {
+        OffHeapTimestampCache offHeapTimestampCache;
+
+        WriteBatcher(OffHeapTimestampCache offHeapTimestampCache) {
+            this.offHeapTimestampCache = offHeapTimestampCache;
+        }
+
+        @Override
+        public Map<Map.Entry<Long, Long>, Map.Entry<Long, Long>> apply(Set<Map.Entry<Long, Long>> request) {
+            if (offHeapTimestampCache.cacheDescriptor.get().currentSize().get() >= offHeapTimestampCache.maxSize) {
+                offHeapTimestampCache.clear();
+            }
+            CacheDescriptor cacheDescriptor = offHeapTimestampCache.cacheDescriptor.get();
+            try {
+                Map<Long, Long> response = offHeapTimestampCache.persistentTimestampStore.multiGet(
+                        cacheDescriptor.storeNamespace(),
+                        request.stream().map(Map.Entry::getKey).collect(Collectors.toList()));
+
+                Set<Map.Entry<Long, Long>> toWrite = Sets.difference(request, response.entrySet());
+                offHeapTimestampCache.persistentTimestampStore.multiPut(
+                        cacheDescriptor.storeNamespace(),
+                        toWrite);
+
+                cacheDescriptor.currentSize().addAndGet(toWrite.size());
+            } catch (SafeIllegalArgumentException exception) {
+                log.warn("Clear called concurrently, writing failed");
+            } finally {
+                offHeapTimestampCache.concurrentHashMap.clear();
```

Comment:
possible race condition here in that `concurrentHashMap` may have extra entries cleared, but this is OK as this is only used to skip reads for values that are about to be written